spynnaker/pyNN/models/neuron/synaptic_manager.py

```diff
@@ -591,120 +605,575 @@ def _write_synaptic_matrix_and_master_population_table(
         # Track writes inside the synaptic matrix region:
         block_addr = 0
 
-        # Get the edges
-        in_edges = machine_graph.get_edges_ending_at_vertex(machine_vertex)
+        # Get the application projection edges incoming to this machine vertex
+        in_machine_edges = machine_graph.get_edges_ending_at_vertex(
+            machine_vertex)
+        in_edges_by_app_edge = defaultdict(list)
+        key_space_tracker = KeySpaceTracker()
+        for edge in in_machine_edges:
+            rinfo = routing_info.get_routing_info_for_edge(edge)
+            key_space_tracker.allocate_keys(rinfo)
+            app_edge = graph_mapper.get_application_edge(edge)
+            if isinstance(app_edge, ProjectionApplicationEdge):
+                in_edges_by_app_edge[app_edge].append(edge)
 
         # Set up the master population table
         self.__poptable_type.initialise_table()
 
         # Set up for single synapses - write the offset of the single synapses
-        # initially 0
-        single_synapses = list()
-        spec.switch_write_focus(synaptic_matrix_region)
+        # initially 0. The list is seeded with an empty array so we can just
+        # concatenate the right blindly.
+        single_synapses = [numpy.array([], dtype="uint32")]
```

Comment:
I think I understand what's going on but the comment could be clearer.

Reply:
OK, made it more comprehensible